### PR TITLE
improve sdk

### PR DESCRIPTION
**Pull request by andrewsource147**

### CODE REVIEW

File: rust-client/src/strategy_handler/mango.rs [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+    fn withdraw_directly_from_strategy(
+        &self,
+        program_client: &anchor_client::Program,
+        strategy: Pubkey,
+        token_mint: Pubkey,
+        base: Pubkey,
+        amount: u64,
+    ) -> Result<()> {
+        let (vault, _vault_bump) = Pubkey::find_program_address(
+            &[b"vault".as_ref(), token_mint.as_ref(), base.as_ref()],
```
Comment (codewithgun) [inline rust-client/src/strategy_handler/mango.rs:26]: We can use the prefix defined in `seed.rs`

File: rust-client/src/strategy_handler/mango.rs [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+            &program_client.id(),
+        );
+
+        let (token_vault, _token_vault_bump) = Pubkey::find_program_address(
+            &[b"token_vault".as_ref(), vault.as_ref()],
```
Comment (codewithgun) [inline rust-client/src/strategy_handler/mango.rs:31]: Ditto^

File: rust-client/src/strategy_handler/port_finance_without_lm.rs [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    fn withdraw_directly_from_strategy(
+        &self,
+        program_client: &anchor_client::Program,
+        strategy: Pubkey,
+        token_mint: Pubkey,
+        base: Pubkey,
+        amount: u64,
+    ) -> Result<()> {
+        let (vault, _vault_bump) = Pubkey::find_program_address(
+            &[b"vault".as_ref(), token_mint.as_ref(), base.as_ref()],
```
Comment (codewithgun) [inline rust-client/src/strategy_handler/port_finance_without_lm.rs:24]: Ditto^

File: rust-client/src/strategy_handler/solend_with_lm.rs [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+    fn withdraw_directly_from_strategy(
+        &self,
+        program_client: &anchor_client::Program,
+        strategy: Pubkey,
+        token_mint: Pubkey,
+        base: Pubkey,
+        amount: u64,
+    ) -> Result<()> {
+        let (vault, _vault_bump) = Pubkey::find_program_address(
+            &[b"vault".as_ref(), token_mint.as_ref(), base.as_ref()],
```
Comment (codewithgun) [inline rust-client/src/strategy_handler/solend_with_lm.rs:25]: Ditto^

File: rust-client/src/strategy_handler/solend_without_lm.rs [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+    fn withdraw_directly_from_strategy(
+        &self,
+        program_client: &anchor_client::Program,
+        strategy: Pubkey,
+        token_mint: Pubkey,
+        base: Pubkey,
+        amount: u64,
+    ) -> Result<()> {
+        let (vault, _vault_bump) = Pubkey::find_program_address(
+            &[b"vault".as_ref(), token_mint.as_ref(), base.as_ref()],
```
Comment (codewithgun) [inline rust-client/src/strategy_handler/solend_without_lm.rs:24]: Ditto^

File: rust-client/src/main.rs [hunk 6]
```diff
@@ -51,8 +48,16 @@ pub enum Command {
...
-    Withdraw { unmint_amount: u64 },
+    Deposit {
+        token_amount: u64,
+    },
+    Withdraw {
+        unmint_amount: u64,
+    },
+    WithdrawFromStrategy {
+        unmint_amount: u64,
+        strategy: Pubkey,
```
Comment (codewithgun) [inline rust-client/src/main.rs:59]: Should we put a mapping of protocol to strategy pubkey at the README ?

Comment (andrewsource147) [reply rust-client/src/main.rs:59]: User can read strategy pubkey through vault state

File: rust-client/src/strategy_handler/mango.rs (hunk 1 above)
Comment (andrewsource147) [reply rust-client/src/strategy_handler/mango.rs:26]: fixed

File: rust-client/src/strategy_handler/mango.rs (hunk 2 above)
Comment (andrewsource147) [reply rust-client/src/strategy_handler/mango.rs:31]: fixed

File: rust-client/src/strategy_handler/port_finance_without_lm.rs (hunk 3 above)
Comment (andrewsource147) [reply rust-client/src/strategy_handler/port_finance_without_lm.rs:24]: fixed

File: rust-client/src/strategy_handler/solend_with_lm.rs (hunk 4 above)
Comment (andrewsource147) [reply rust-client/src/strategy_handler/solend_with_lm.rs:25]: fixed

**Review by codewithgun — APPROVED** [review]

**Pull request merged**